LoopKit/GlucoseKit/GlucoseStore.swift

```diff
@@ -143,511 +150,496 @@ public final class GlucoseStore: HealthKitSampleStore {
+        let quantitySamples = samples.map { $0.quantitySample }
+        var error: Error?
 
-        var created = false
-        var storageError: Error? = nil
-        
-        cacheStore.managedObjectContext.performAndWait {
-            for sample in samples {
-                guard
-                    sample.startDate.timeIntervalSinceNow > -self.cacheLength,
-                    self.cacheStore.managedObjectContext.cachedGlucoseObjectsWithUUID(sample.sampleUUID, fetchLimit: 1).count == 0
-                else {
-                    continue
-                }
+        // Save objects to HealthKit, log any errors, but do not fail
+        let dispatchGroup = DispatchGroup()
+        dispatchGroup.enter()
+        self.healthStore.save(quantitySamples) { (_, healthKitError) in
+            error = healthKitError
+            dispatchGroup.leave()
+        }
+        dispatchGroup.wait()
 
-                let object = CachedGlucoseObject(context: self.cacheStore.managedObjectContext)
-                object.update(from: sample)
-                created = true
-            }
+        if let error = error {
+            self.log.error("Error saving HealthKit objects: %@", String(describing: error))
+            return
+        }
 
-            if created {
-                self.cacheStore.save { (error) in
-                    storageError = error
-                }
-            }
+        // Update Core Data with the changes, log any errors, but do not fail
+        for (object, quantitySample) in zip(objects, quantitySamples) {
+            object.uuid = quantitySample.uuid
         }
-        
-        if let error = storageError {
-            return .failure(error)
-        } else {
-            return .success(created)
+        if let error = self.cacheStore.save() {
+            self.log.error("Error updating CachedGlucoseObjects after saving HealthKit objects: %@", String(describing: error))
+            objects.forEach { $0.uuid = nil }
         }
     }
 
-    private func getCachedGlucoseObjects(start: Date, end: Date? = nil) -> [StoredGlucoseSample] {
-        dispatchPrecondition(condition: .onQueue(dataAccessQueue))
-        let predicate: NSPredicate
+    private func addGlucoseSample(for sample: HKQuantitySample) throws -> Bool {
+        dispatchPrecondition(condition: .onQueue(queue))
 
-        if let end = end {
-            predicate = NSPredicate(format: "startDate >= %@ AND startDate <= %@", start as NSDate, end as NSDate)
-        } else {
-            predicate = NSPredicate(format: "startDate >= %@", start as NSDate)
+        // Are there any objects matching the UUID?
+        let request: NSFetchRequest<CachedGlucoseObject> = CachedGlucoseObject.fetchRequest()
+        request.predicate = NSPredicate(format: "uuid == %@", sample.uuid as NSUUID)
+        request.fetchLimit = 1
+
+        let count = try cacheStore.managedObjectContext.count(for: request)
+        guard count == 0 else {
+            return false
```

Comment:
[nit] Log this unexpected state? Or is it expected at times.

Reply:
Added log in calling code.